File: services/nlp/app/services/content/gutenberg.py
"""
Project Gutenberg content discovery (via the Gutendex catalog API).

Python port of the PHP ``GutenbergClient``
(``src/Shared/Infrastructure/Http/GutenbergClient.php``). Searches the Gutendex
catalog (``gutendex.com``) for free public-domain e-books, normalises each book
record, and resolves a plain-text download URL. ``fetch_text`` downloads a book
and strips the standard Project Gutenberg header/footer boilerplate.

All outbound requests go through the SSRF-safe fetch util so every hop (including
redirects out of gutendex.com) is re-validated.
"""

from __future__ import annotations

import re

from app.config import settings
from app.services.http.safe_fetch import safe_get, safe_get_json

# Matches Project Gutenberg's start/end boilerplate markers, tolerant of the
# "THE"/"THIS" wording variants and any trailing book title text, e.g.:
#   *** START OF THE PROJECT GUTENBERG EBOOK MOBY DICK ***
#   *** END OF THIS PROJECT GUTENBERG EBOOK ... ***
_START_MARKER = re.compile(
    r"\*\*\*\s*START OF (?:THE|THIS) PROJECT GUTENBERG EBOOK.*?\*\*\*",
    re.IGNORECASE | re.DOTALL,
)
_END_MARKER = re.compile(
    r"\*\*\*\s*END OF (?:THE|THIS) PROJECT GUTENBERG EBOOK.*?\*\*\*",
    re.IGNORECASE | re.DOTALL,
)


def _extract_text_url(book: dict) -> str:
    """Pick the best plain-text download URL from a Gutendex book record.

    Prefers an explicitly UTF-8 / ``.txt`` plain-text format, falling back to
    any ``text/plain`` entry. Returns "" when the book has no plain text.
    """
    formats = book.get("formats") or {}
    if not isinstance(formats, dict):
        return ""

    plain_entries = [
        (mime, url)
        for mime, url in formats.items()
        if isinstance(mime, str) and "text/plain" in mime and isinstance(url, str)
    ]

    # Prefer an explicitly UTF-8 entry (matches the PHP client's precedence),
    # then a ``.txt`` / "utf-8" URL, then any plain-text entry.
    for mime, url in plain_entries:
        if "utf-8" in mime.lower():
            return url
    for mime, url in plain_entries:
        lower_url = url.lower()
        if lower_url.endswith(".txt") or "utf-8" in lower_url:
            return url
    if plain_entries:
        return plain_entries[0][1]

    return ""


def _map_book(book: dict) -> dict:
    """Normalise a raw Gutendex book record into the API result shape."""
    authors: list[str] = []
    for author in book.get("authors") or []:
        if isinstance(author, dict):
            authors.append(str(author.get("name", "")))

    subjects = book.get("subjects") or []
    if not isinstance(subjects, list):
        subjects = []

    languages = book.get("languages") or []
    if not isinstance(languages, list):
        languages = []

    return {
        "id": int(book.get("id") or 0),
        "title": str(book.get("title") or ""),
        "authors": authors,
        "languages": languages,
        "subjects": [str(s) for s in subjects[:3]],
        "downloadCount": int(book.get("download_count") or 0),
        "textUrl": _extract_text_url(book),
    }


async def search(
    query: str, language_code: str | None = None, page: int = 1
) -> dict:
    """Search the Project Gutenberg catalog via Gutendex.

    Maps each upstream book into the normalised result shape. Top-level keys:
    ``results`` (list), ``count`` (total matches), ``next`` (more pages exist).
    """
    params: dict[str, object] = {}
    if query:
        params["search"] = query
    if language_code:
        params["languages"] = language_code.lower()
    if page > 1:
        params["page"] = page

    response = await safe_get_json(settings.gutendex_base_url, params=params)
    if not isinstance(response, dict):
        response = {}

    raw_results = response.get("results") or []
    results = [_map_book(b) for b in raw_results if isinstance(b, dict)]

    return {
        "results": results,
        "count": int(response.get("count") or 0),
        "next": response.get("next") is not None,
    }


async def browse(language_code: str, page: int = 1) -> dict:
    """Browse popular books for a language (search with an empty query)."""
    return await search("", language_code, page)


async def fetch_text(url: str) -> str:
    """Download a Gutenberg plain-text book and strip its boilerplate.

    Everything before the ``*** START OF ... ***`` marker and after the
    ``*** END OF ... ***`` marker is removed. When the markers are absent the
    full text is returned unchanged.
    """
    result = await safe_get(
        url, timeout=15.0, max_bytes=2_000_000, accept="text/plain"
    )
    text = result.text

    start = _START_MARKER.search(text)
    if start:
        text = text[start.end():]

    end = _END_MARKER.search(text)
    if end:
        text = text[: end.start()]

    return text.strip()

File: src/frontend/js/shared/offline/local/content/gutenberg.ts
/**
 * Project Gutenberg catalog client (Gutendex), ported to run client-side.
 *
 * Mirrors the server's `GutenbergClient` + `services/nlp/.../content/gutenberg.py`:
 * browse/search via the Gutendex API, pick the best UTF-8 plain-text URL from a
 * book's `formats` map, and strip the `*** START/END OF ... PROJECT GUTENBERG
 * EBOOK ***` boilerplate when importing. All network access goes through the
 * CORS-free helper so it works inside the packaged app with no server.
 *
 * @license Unlicense <http://unlicense.org/>
 */

import { corsFreeGet, corsFreeGetJson } from './native-http';
import type { DifficultyTier } from './difficulty';

const GUTENDEX_URL = 'https://gutendex.com/books/';

/** A Gutenberg book as the suggestion UI consumes it. */
export interface GutenbergBook {
  id: number;
  title: string;
  authors: string[];
  languages: string[];
  subjects: string[];
  downloadCount: number;
  textUrl: string;
  difficultyTier?: DifficultyTier;
}

/** A page of catalog results, matching the server's `{ results, count, next }`. */
export interface CatalogPage<T> {
  results: T[];
  count: number;
  next: boolean;
}

/** Raw Gutendex book record (only the fields we read). */
interface GutendexBook {
  id?: number;
  title?: string;
  authors?: Array<{ name?: string }>;
  languages?: string[];
  subjects?: string[];
  download_count?: number;
  formats?: Record<string, string>;
}

interface GutendexResponse {
  count?: number;
  next?: string | null;
  results?: GutendexBook[];
}

/**
 * Pick the best plain-text download URL from a Gutendex `formats` map. Prefers
 * an explicitly UTF-8 entry, then a `.txt`/"utf-8" URL, then any `text/plain`.
 * Mirrors `gutenberg.py::_extract_text_url`.
 */
export function extractTextUrl(formats: Record<string, string> | undefined): string {
  if (!formats || typeof formats !== 'object') {
    return '';
  }
  const plain: Array<[string, string]> = [];
  for (const [mime, url] of Object.entries(formats)) {
    if (typeof mime === 'string' && mime.includes('text/plain') && typeof url === 'string') {
      plain.push([mime, url]);
    }
  }
  for (const [mime, url] of plain) {
    if (mime.toLowerCase().includes('utf-8')) {
      return url;
    }
  }
  for (const [, url] of plain) {
    const lower = url.toLowerCase();
    if (lower.endsWith('.txt') || lower.includes('utf-8')) {
      return url;
    }
  }
  return plain.length > 0 ? plain[0][1] : '';
}

function mapBook(raw: GutendexBook): GutenbergBook {
  return {
    id: raw.id ?? 0,
    title: raw.title ?? '',
    authors: (raw.authors ?? []).map((a) => a.name ?? '').filter((n) => n !== ''),
    languages: raw.languages ?? [],
    subjects: (raw.subjects ?? []).slice(0, 3),
    downloadCount: raw.download_count ?? 0,
    textUrl: extractTextUrl(raw.formats),
  };
}

/**
 * Browse (empty query) or search the Gutendex catalog. `languageCode` is an ISO
 * 639-1 code or null for no filter; `page` is 1-based.
 */
export async function searchGutenberg(
  query: string,
  languageCode: string | null,
  page: number
): Promise<CatalogPage<GutenbergBook>> {
  const params = new URLSearchParams();
  if (query) {
    params.set('search', query);
  }
  if (languageCode) {
    params.set('languages', languageCode);
  }
  if (page > 1) {
    params.set('page', String(page));
  }
  const qs = params.toString();
  const data = await corsFreeGetJson<GutendexResponse>(GUTENDEX_URL + (qs ? `?${qs}` : ''));
  return {
    results: (data.results ?? []).map(mapBook),
    count: data.count ?? 0,
    next: Boolean(data.next),
  };
}

// Everything before the START marker and after the END marker is boilerplate.
// `[\s\S]*?` instead of the `s`-flag so the regex works regardless of target.
const START_MARKER = /\*\*\*\s*START OF (?:THE|THIS) PROJECT GUTENBERG EBOOK[\s\S]*?\*\*\*/i;
const END_MARKER = /\*\*\*\s*END OF (?:THE|THIS) PROJECT GUTENBERG EBOOK[\s\S]*?\*\*\*/i;

/**
 * Strip Project Gutenberg header/footer boilerplate. When the markers are
 * absent the text is returned trimmed but otherwise unchanged. Mirrors
 * `gutenberg.py::fetch_text`'s stripping logic.
 */
export function stripGutenbergBoilerplate(text: string): string {
  let body = text;
  const start = START_MARKER.exec(body);
  if (start) {
    body = body.slice(start.index + start[0].length);
  }
  const end = END_MARKER.exec(body);
  if (end) {
    body = body.slice(0, end.index);
  }
  return body.trim();
}

/**
 * Download a Gutenberg plain-text book and strip its boilerplate. Returns ''
 * when the book is unreachable or empty.
 */
export async function fetchGutenbergText(url: string): Promise<string> {
  const res = await corsFreeGet(url, { accept: 'text/plain', timeoutMs: 20000 });
  if (!res.ok || !res.text) {
    return '';
  }
  return stripGutenbergBoilerplate(res.text);
}
